### ISSUE TITLE
How to concatenate parts of URI?

### ISSUE THREAD

**Issue by ChShersh** [mrkkrp/modern-uri#10]
I wonder if there some concatenating operators for parts of complex URI? Similar to `</>` from `filepath` library. For example, I want to have:

1. `https://markkarpov.com`
2. `https://markkarpov.com/foo`
3. `https://markkarpov.com/bar`
4. `https://markkarpov.com/bar/baz`

Sure, I can write four hardcoded quasi quotes. But I wonder what is preferred and simplest way to do this reusing first URI?

I found [`mkPathPiece`](https://hackage.haskell.org/package/modern-uri-0.2.0.0/docs/Text-URI.html#v:mkPathPiece) smart constructor but have no idea how to reuse it...

**Comment by mrkkrp** [comment]
Concatenation by itself is only good for ordered collections with regular structure such as lists. Even file paths are more complex than just a list of path components, because they may be absolute or relative (even if we ignore drive letters on Windows).

This is obvious if we recall that the `(</>)` operation from `filepath` is not quite concatenation, it has to be more complex because it has to deal with cases like when its right (second) argument is absolute. What to do in that case? We need to adopt a convention: we just return the second argument unchanged in that case.

URIs are even more complex than file paths, they have many different components and merging two URIs (I don't even say "concatenate") should also be done by using a convention. Luckily the standard describing URIs, RFC 3986 [describes how to resolve one URI against another *absolute* URI](https://tools.ietf.org/html/rfc3986#section-5), which you can do with `modern-uri-0.2.0.0` and later, see [`relativeTo`](https://hackage.haskell.org/package/modern-uri-0.2.0.0/docs/Text-URI.html#v:relativeTo).

If you just want to operate on path pieces only, you first need to obtain the path pieces. Since not every `Text` is a valid path piece, you must prove that your candidate is a valid path piece by feeding it into `mkPathPiece` and obtaining a refined text value (which serves as a proof that that `Text` value is indeed a valid path piece). This "proving" may conclude that your candidate is not a valid path piece, of course, that's why we have the `MonadThrow` constraint. If you click on `MonadThrow` you'll see that it has pure instances like `Maybe` too, so you can work in that context.

When you have your path pieces, you may do whatever you want by using lens:

```haskell
λ> :set -XOverloadedStrings
λ> uri <- URI.mkURI "https://markkarpov.com"
λ> pp <- URI.mkPathPiece "foo"
λ> URI.render (uri & uriPath %~ (pp:))
"https://markkarpov.com/foo"
```

**Comment by ChShersh** [comment]
@mrkkrp Thanks for your response! Sure, I don't want to concatenate two URIs. Using lenses is good, I just wasn't sure how to do it exactly. I would appreciate having some combinator like this:

```haskell
infixl 7 </>  -- not sure about precedence
(</>) :: URI -> RText PathPiece -> URI
uri </> pp = uri & uriPath %~ (++ [pp])
```

**Comment by mrkkrp** [comment]
I think such a combinator is not a good idea, because from the design point of view we step into the area of "too specific" combinations of existing building blocks. My philosophy is to provide the basic building blocks, which are simple, few, and compose well and let the user combine them to solve his/her specific tasks. In addition to that, such a function would make it tempting to use it repeatedly to build a path, and given the implementation appending will have quadratic complexity.

**Comment by ChShersh** [comment]
> and given the implementation appending will have quadratic complexity.

Yes, this is sad, I agree. Probably, switching from list to `Seq` from `Data.Sequence` can solve performance problem. But I'm not sure is it worth it. I agree with your vision on providing basic building blocks. Your answer already helped me and resolved my issue!

**Comment by ChShersh** [comment]
Btw, `uri & uriPath %~ (pp:)` doesn't really work well because since version `0.2.0.0`  `uriPath :: Maybe (Bool, NonEmpty (RText PathPiece))`.

**Comment by mrkkrp** [comment]
Not really, `uriPath`, the lens, still manipulates plain list of path pieces:

https://hackage.haskell.org/package/modern-uri-0.2.0.0/docs/Text-URI-Lens.html#v:uriPath

For your convenience!

**Comment by ChShersh** [comment]
Oh, I see. Thanks a lot!
